### PR TITLE
Implement study staff spreadsheet handler and tests

### LINKED ISSUE #5: Add google spreadsheet handler and cover it with tests

### PR DESCRIPTION
**Pull request by MksmOrlov**

### CODE REVIEW

File: sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+from bot.modules.spreadsheet.spreadsheet_handler import SpreadsheetHandler
+
+
+class StudyStaffSpreadsheetHandler:
+    def __init__(self, spreadsheet_id: str, file_name: str):
+        self._attributes = {
+            "Студенты": ["username", "ФИО", "Группа", "Подгруппа"],
+            "Преподаватели": ["username", "ФИО"],
+        }
```
Comment (NikitaZotov) [inline sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py]: It's cool to make attributes static.

File: sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        self._handler = SpreadsheetHandler(spreadsheet_id, file_name, self._attributes)
+        self._student_sheet_title = list(self._attributes.keys())[0]
+        self._teacher_sheet_title = list(self._attributes.keys())[1]
+
+    def create_spreadsheet(self, spreadsheet_title="Информация о людях", row_count=1000, column_count=10) -> None:
+        self._handler.create_spreadsheet(spreadsheet_title, row_count, column_count)
+
+    def add_student(self, username: str, name: str, group: str, subgroup: str) -> None:
```
Comment (NikitaZotov) [inline sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py]: Why are we passing so much args? To make a comfortable interface, it's better to pass **kwargs or *args. It's so add_teacher

File: sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+from typing import List, Dict
+
+from bot.modules.spreadsheet.spreadsheet_handler import SpreadsheetHandler
+
+
+class StudyStaffSpreadsheetHandler:
```
Comment (NikitaZotov) [inline sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py]: I think we need to implement an interface for this class

File: sources/bot/modules/spreadsheet/study_staff/tests/spreadsheet_test.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import unittest
+
+from bot.modules.spreadsheet.study_staff.study_staff_spreadsheet_handler import StudyStaffSpreadsheetHandler
+
+
+class TestSpreadsheet(unittest.TestCase):
+    def setUp(self):
+        token_path = "../../spreadsheet_token.json"
+        self.handler = StudyStaffSpreadsheetHandler("", token_path)
+        self.handler.create_spreadsheet()
```
Comment (NikitaZotov) [inline sources/bot/modules/spreadsheet/study_staff/tests/spreadsheet_test.py:10]: It's cool job

File: sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py (hunk 2 above)
Comment (NikitaZotov) [reply sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py]: Let's speak about it

**Review by NikitaZotov — CHANGES_REQUESTED** [review]

File: sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py (hunk 1 above)
Comment (MksmOrlov) [reply sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py]: fixed

File: sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py (hunk 2 above)
Comment (MksmOrlov) [reply sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py]: If pass **kwargs or *args you whouldnt see how much and what concrete paramentr need to pass using this class as API

File: sources/bot/modules/spreadsheet/study_staff/tests/spreadsheet_test.py (hunk 4 above)
Comment (MksmOrlov) [reply sources/bot/modules/spreadsheet/study_staff/tests/spreadsheet_test.py:10]: ### Ok

File: sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py (hunk 2 above)
Comment (NikitaZotov) [reply sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py]: They need to be passed in order to make this class easy to replace by others across polymorphism without stopping at these parameters.

Comment (MksmOrlov) [reply sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py]: fixed

File: sources/bot/modules/spreadsheet/study_staff/tests/students_test.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        self.handler.add_student("student2", "name2", "group2", "subgroup2")
+        self.handler.add_student("student3", "name3", "group3", "subgroup3")
+
+        student2 = self.handler.get_student_by_username("student2")
+        student_usernames = self.handler.get_student_usernames()
+
+        self.assertEqual(student2.get("username"), "student2")
+        self.assertEqual(len(student_usernames), 3)
+
+        self.handler.remove_student("student2")
```
Comment (NikitaZotov) [inline sources/bot/modules/spreadsheet/study_staff/tests/students_test.py]: What does this method return if it won't find student with such username?

File: sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        if not name:
+            raise InvalidSpreadsheetAttributeException("Invalid name value")
+        elif not group:
+            raise InvalidSpreadsheetAttributeException("Invalid group value")
+        elif not subgroup:
+            raise InvalidSpreadsheetAttributeException("Invalid subgroup value")
+        else:
+            self._handler.add_row(self._student_sheet_title, [username, name, group, subgroup])
+
+    def remove_student(self, username: str) -> None:
```
Comment (NikitaZotov) [inline sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py]: It is needed that this method return boolean value

File: sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+    def add_teacher(self, username: str, *args) -> None:
+        name = args[0]
+
+        if not name:
+            raise InvalidSpreadsheetAttributeException("Invalid name value")
+        else:
+            self._handler.add_row(self._teacher_sheet_title, [username, name])
+
+    def remove_teacher(self, username: str) -> None:
```
Comment (NikitaZotov) [inline sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py]: It's so

**Review by NikitaZotov — CHANGES_REQUESTED** [review]

File: sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py (hunk 3 above)
Comment (MksmOrlov) [reply sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py]: done

File: sources/bot/modules/spreadsheet/study_staff/tests/students_test.py (hunk 5 above)
Comment (MksmOrlov) [reply sources/bot/modules/spreadsheet/study_staff/tests/students_test.py]: bool value. fixed

File: sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py (hunk 6 above)
Comment (MksmOrlov) [reply sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py]: fixed

File: sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py (hunk 7 above)
Comment (MksmOrlov) [reply sources/bot/modules/spreadsheet/study_staff/study_staff_spreadsheet_handler.py]: fixed so

File: sources/bot/modules/spreadsheet/study_staff/base_study_staff_spreadsheet_handler.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    @abstractmethod
+    def create_spreadsheet(self, spreadsheet_title, row_count, column_count) -> None:
+        raise SpreadsheetHandlerException("Not implemented method")
+
+    @abstractmethod
+    def add_student(self, username: str, *args) -> None:
+        raise SpreadsheetHandlerException("Not implemented method")
+
+    @abstractmethod
+    def remove_student(self, username: str) -> bool:
```
Comment (NikitaZotov) [inline sources/bot/modules/spreadsheet/study_staff/base_study_staff_spreadsheet_handler.py:19]: Sort methods in order

File: sources/bot/modules/spreadsheet/study_staff/tests/teacher_test.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        self.handler.add_teacher("teacher3", "name3")
+
+        teacher2 = self.handler.get_teacher_by_username("teacher2")
+        teacher_usernames = self.handler.get_teacher_usernames()
+
+        self.assertEqual(teacher2.get("username"), "teacher2")
+        self.assertEqual(len(teacher_usernames), 3)
+
+        self.assertFalse(self.handler.remove_teacher("unknown"))
+        self.assertTrue(self.handler.remove_teacher("teacher2"))
```
Comment (NikitaZotov) [inline sources/bot/modules/spreadsheet/study_staff/tests/teacher_test.py:46]: Good

**Review by NikitaZotov — CHANGES_REQUESTED** [review]

File: sources/bot/modules/spreadsheet/study_staff/tests/teacher_test.py (hunk 9 above)
Comment (MksmOrlov) [reply sources/bot/modules/spreadsheet/study_staff/tests/teacher_test.py:46]: ### Ok

File: sources/bot/modules/spreadsheet/study_staff/base_study_staff_spreadsheet_handler.py (hunk 8 above)
Comment (MksmOrlov) [reply sources/bot/modules/spreadsheet/study_staff/base_study_staff_spreadsheet_handler.py:19]: They are already sorted

**Pull request merged**